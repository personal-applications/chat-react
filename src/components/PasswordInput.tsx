import React, { useState } from "react";
import Input, { InputProps } from "./Input";
import { Else, If, Then } from "react-if";
import IconEye from "./icons/IconEye";
import IconEyeInvisible from "./icons/IconEyeInvisible";

export type PasswordInputProps = Omit<InputProps, "type">;
export type PasswordInputType = "password" | "text";

const PasswordInput = React.forwardRef(
  (props: PasswordInputProps, ref: React.Ref<HTMLInputElement>) => {
    const [inputType, setInputType] = useState<PasswordInputType>("password");

    const onToggleInputType = () => {
      setInputType((prev) => (prev === "password" ? "text" : "password"));
    };

    return (
      <>
        <Input
          {...props}
          type={inputType}
          ref={ref}
          postIcon={
            <button type="button" onClick={onToggleInputType}>
              <If condition={inputType === "password"}>
                <Then>
                  <IconEye />
                </Then>
                <Else>
                  <IconEyeInvisible />
                </Else>
              </If>
            </button>
          }
        />
      </>
    );
  },
);

export default PasswordInput;
