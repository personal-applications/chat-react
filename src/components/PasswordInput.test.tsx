import { fireEvent, render, screen } from "@testing-library/react";
import PasswordInput from "./PasswordInput";

test("renders without crashing", () => {
  const tree = render(<PasswordInput placeholder="Enter password" />);

  expect(tree).toMatchSnapshot();
});

test("toggles password visibility", () => {
  render(<PasswordInput placeholder="Enter password" />);
  const input = screen.getByPlaceholderText("Enter password");
  const toggleButton = screen.getByRole("button");

  // Initially, the input type should be 'password'
  expect(input).toHaveAttribute("type", "password");

  // After clicking the toggle button, the input type should be 'text'
  fireEvent.click(toggleButton);
  expect(input).toHaveAttribute("type", "text");

  // After clicking the toggle button again, the input type should be 'password' again
  fireEvent.click(toggleButton);
  expect(input).toHaveAttribute("type", "password");
});
