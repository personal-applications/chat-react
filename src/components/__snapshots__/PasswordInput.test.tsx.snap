// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders without crashing 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="input input-bordered flex items-center gap-2"
        >
          <input
            class="grow"
            placeholder="Enter password"
            type="password"
          />
          <button
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M396 512a112 112 0 10224 0 112 112 0 10-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
              />
            </svg>
          </button>
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex w-full flex-col"
    >
      <label
        class="input input-bordered flex items-center gap-2"
      >
        <input
          class="grow"
          placeholder="Enter password"
          type="password"
        />
        <button
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M396 512a112 112 0 10224 0 112 112 0 10-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
            />
          </svg>
        </button>
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
